src/qibo/models/double_bracket.py

```diff
@@ -95,3 +97,65 @@ def off_diagonal_norm(self):
     def backend(self):
         """Get Hamiltonian's backend."""
         return self.h0.backend
+
+    def hyperopt_step(
+        self,
+        step_min: float = 0.0001,
+        step_max: float = 0.5,
+        max_evals: int = 1000,
+        space: callable = hyperopt.hp.uniform,
+        optimizer: callable = hyperopt.tpe,
+        look_ahead: int = 1,
+        verbose: bool = False,
+    ):
+        """
+        Optimize flow step.
+
+        Args:
+            step_min: lower bound of the search grid;
+            step_max: upper bound of the search grid;
+            max_evals: maximum number of iterations done by the hyperoptimizer;
+            space: see hyperopt.hp possibilities;
+            optimizer: see hyperopt algorithms;
+            look_ahead: number of flow steps to compute the loss function;
+            verbose: level of verbosity.
+
+        Returns:
+            (float): optimized best flow step.
+        """
+
+        space = space("step", step_min, step_max)
+        best = hyperopt.fmin(
+            fn=partial(self.loss, look_ahead=look_ahead),
+            space=space,
+            algo=optimizer.suggest,
+            max_evals=max_evals,
+            verbose=verbose,
+        )
+
+        return best["step"]
+
+    def loss(self, step: float, look_ahead: int = 1):
+        """
+        Compute loss function distance between `look_ahead` steps.
+
+        Args:
+            step: flow step.
+            look_ahead: number of flow steps to compute the loss function;
+        """
+        # copy initial hamiltonian
+        h_copy = self.h
+
+        # off_diagonal_norm's value before the steps
+        old_loss = self.off_diagonal_norm
+
+        for _ in range(look_ahead):
+            self.__call__(mode=FlowGeneratorType.canonical, step=step)
```

Comment:
Here you are assuming that the `FlowGeneratorType` is `canonical`.
In theory it could be any of the types present in the `Enum`.
Perhaps the easiest way would be to pass `FlowGeneratorType` in the constructor.

Reply:
I agree, I think it is wise to set it while initializing the model.